source/views/components/nav-buttons/favorite.js

```diff
@@ -0,0 +1,55 @@
+// @flow
+
+import * as React from 'react'
+import {Platform} from 'react-native'
+import Icon from 'react-native-vector-icons/Ionicons'
+import {Touchable} from '../touchable'
+import {rightButtonStyles as styles} from './styles'
+
+type State = {
+  favorited: boolean,
+}
+
+type Props = {
+  onFavorite: () => any,
+  favorited: boolean,
+}
+
+export class FavoriteButton extends React.PureComponent<Props, State> {
+  state = {
```

Comment:
When you move the logic of saving the favorite into the redux file and make the "am I favorited" logic on the Detail screen be a (mathematical) function of the Redux state, this component won't need any internal state any longer; the lifecycle will go something like this:

- Person taps "favorite"
- Redux "action creator" onSaveFavorites is called
- Said function determines an array of actually favorited buildings (and saves it to disk)
- Created action is dispatched via the `connect`ed dispatch function
- Redux state updates with new array of favorites building names
- Redux state update causes React to re-render the components that have connected to the store
- StatefulList gets the new set of building names and groups them again
- DetailView gets the new set of building names and figures out if it's favorited
- If it is, it sets the "favorited" prop of FavoriteButton to "true" or "false", and react calls FavoriteButton's render function again with the new props (otherwise nothing happens)

At that point, there's no reason to keep the separate state in this button, so it becomes a much smaller component. (Not that it's not already small.)

If you try that and the button doesn't seem to be working, please let me know - the router that handles new screens might be interfering with the props getting passed right (it may not be passing the new ones, since navigate is only called once and we haven't changed screens). Or it may not! I'm not sure.

Reply:
Ok, I have implemented the changes you asked for, but I am unsure how to get the navigationOptions in the detail view to refresh when the view is re-rendered by redux upon the state change. The only way I can get the button to update properly is if I pass navigation to the onSaveFavorites and saveFavoriteBuildings functions, respectively and then call navigation.setParams({favoriteBuildings: favoriteBuildings}) within saveFavoriteBuildings in buildings.js...is there a way to get the navigationOptions to automatically refresh when the redux state changes?

Reply:
@hawkrives Regarding "npm run check"..running yarn does not make the "check" script work..

Reply:
Sigh… that's what I was afraid of, that the toolbar buttons wouldn't be able to subscribe to redux. 

I'll think about this and probably get back to you tomorrow … I'm thinking that it might be possible / simple to connect to redux on the FavoriteButton itself, and pass the building name to the button, then use the buildings name in mapStateToProps within the button's redux connection to determine if the building is connected or not. 

I'll try to work up something for that approach tomorrow.

Reply:
Sounds good. I just pushed the latest version that implemented most of the changes you requested.